server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -469,10 +503,17 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
             // now, go over all the shards routing on the node, and fail them
             for (ShardRouting shardRouting : node.copyShards()) {
                 final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
-                boolean delayed = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0;
+                boolean delayedDueToKnownRestart = Optional.ofNullable(nodesShutdownMetadata.get(node.nodeId()))
+                    .map(shutdown -> Type.RESTART.equals(shutdown.getType()) && shutdown.getShardReallocationDelay().nanos() > 0)
+                    .orElse(false);
+                boolean delayed = delayedDueToKnownRestart
+                    || INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0;
+                boolean nodeIsRestarting = Optional.ofNullable(nodesShutdownMetadata.get(shardRouting.currentNodeId()))
```

Comment:
I think we should always have `delayedDueToKnownRestart == nodeIsRestarting` so we can hopefully do just one of these calculations?

Reply:
This isn't 100% resolved - we still have to decide if the `reason` is `NODE_LEFT` or `NODE_RESTARTING`, which is separate from the `delayedDueToKnownRestart` because `delayed` should be `false` even if the node is restarting if the delay would be zero (as you pointed out in your first review).